manticore/core/smtlib/solver.py

```diff
@@ -143,207 +167,162 @@ def minmax(self, constraints, x, iters=10000):
-                # z3 was too fast to close
-                logger.debug(str(e))
-            finally:
-                try:
-                    self._proc.stdin.close()
-                except IOError as e:
-                    logger.debug(str(e))
-                try:
-                    self._proc.stdout.close()
-                except IOError as e:
-                    logger.debug(str(e))
-                self._proc.kill()
-                # Wait for termination, to avoid zombies.
-                self._proc.wait()
-
-        self._proc: Popen = None
-
-    # marshaling/pickle
-    def __getstate__(self):
-        raise SolverException()
-
-    def __setstate__(self, state):
-        raise SolverException()
-
-    def __del__(self):
-        try:
-            if self._proc is not None:
-                self._stop_proc()
-            # self._proc.stdin.writelines(('(exit)\n',))
-            # self._proc.wait()
-        except Exception as e:
-            logger.error(str(e))
-
-    def _reset(self, constraints: Optional[str] = None) -> None:
-        """Auxiliary method to reset the smtlib external solver to initial defaults"""
-        if self._proc is None:
-            self._start_proc()
-        else:
-            if self.support_reset:
-                self._send("(reset)")
-
-                for cfg in self._init:
-                    self._send(cfg)
-            else:
-                self._stop_proc()
-                self._start_proc()
-        if constraints is not None:
-            self._send(constraints)
-
-    def _send(self, cmd: str) -> None:
+            self._proc.stdin.close()
+            self._proc.stdout.close()
+            # Kill the process
+            self._proc.kill()
+            # Wait for termination, to avoid zombies.
+            self._proc.wait()
+        self._proc = None
+
+    def __readline_and_count(self):
+        assert self._proc
+        assert self._proc.stdout
+        buf = self._proc.stdout.readline()  # No timeout enforced here
+        # lparen, rparen = buf.count("("), buf.count(")")
+        lparen, rparen = map(sum, zip(*((c == "(", c == ")") for c in buf)))
+        return buf, lparen, rparen
+
+    def send(self, cmd: str) -> None:
         """
         Send a string to the solver.
 
         :param cmd: a SMTLIBv2 command (ex. (check-sat))
         """
-        # logger.debug('>%s', cmd)
-        try:
-            if self._proc.stdout:
-                self._proc.stdout.flush()
-            else:
-                raise SolverError("Could not flush stdout: file descriptor is None")
-            if self._proc.stdin:
-                self._proc.stdin.write(f"{cmd}\n")
-            else:
-                raise SolverError("Could not write to stdin: file descriptor is None")
-        except IOError as e:
-            raise SolverError(str(e))
-
-    def _recv(self) -> str:
-        """Reads the response from the solver"""
+        assert self._proc
```

Comment:
Since these now raise `AssertionError` instead of `SolverError`, we should check our error handling elsewhere to make sure that this won't break uncovered code. At a glance, it looks like the linux module might rely on it.

Reply:
I was hoping not to add an IF here and let it fail by other means in such case.
Then added the asserts to silence mypy.  
So..  
option1: Re add the IF
option2: # ignore: type
option3: leave the asserts

I'm very gently in favor of option2

Reply:
Removed the assertions. Still need to think what's the exception game of solver.py. Should we try to catch and  translate the exceptions to a certain type?